site/book/03-packages/01-getting-a-package.md

```diff
@@ -0,0 +1,118 @@
+  type: git
+  gitLock:
+    repo: https://github.com/GoogleContainerTools/kpt
+    directory: /package-examples/wordpress
+    ref: package-examples/wordpress/v0.1
+    commit: e0e0b3642969c2d14fe1d38d9698a73f18aa848f
+info:
+  emails:
+    - kpt-team@google.com
+  description: This is an example wordpress package with mysql subpackage
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        wp-image: wordpress
+        wp-tag: 4.8-apache
+```
+
+The `Kptfile` contains two sections to keep track of the upstream package:
+
+1. The `upstream` section contains the user-specified reference to the upstream package. This
+   contains three pieces of information:
+   - `repo`: The git repository where the package can be found
+   - `directory`: The directory within the git repository where this package can be found
+   - `ref`: The git reference for the package. This can be either a brach, tag, or commit SHA.
+2. The `upstreamLock` section contains the resolved upstream reference. This section is managed by
+   kpt and should not be changed manually.
+
+So essentially, the `upstream` section defines the "desired state" while the `upstreamLock` section
+defines the “current state”.
+
+Now, let's look at the `Kptfile` for the `mysql` subpackage:
+
+```shell
+$ cat wordpress/mysql/Kptfile
+```
+
+```yaml
+apiVersion: kpt.dev/v1alpha2
+kind: Kptfile
+metadata:
+  name: mysql
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        ms-image: mysql
+        ms-tag: 5.6
+```
+
+As you can see, this `Kptfile` doesn't have the `upstream` and `upstreamLock` sections.
+This is because there are two different package types in kpt:
+
+- **Independent package:** A package where the `Kptfile` has `upstream` defined.
+- **Dependent package:** A package where the `Kptfile` doesn’t have `upstream` defined.
+
+In this case, the `mysql` subpackage is a _dependent package_. The upstream package for `mysql` is
+automatically inferred from the parent package. You can think of the `Kptfile` in the `mysql`
+package as implicitly inheriting the `upstream` section of its parent, with the only difference
+being that `upstream.directory` points to `/package-examples/wordpress/mysql`.
+
+## Package Name and Identifier
+
+It is possible to specify a different local directory name to the `get` command. For example,
+the following fetches the packages to a directory names `mywordpress`:
+
+```shell
+$ kpt pkg get https://github.com/GoogleContainerTools/kpt.git/package-examples/wordpress@v0.1 mywordpress
+```
+
+The _name_ of a package is given by its directory name. Since the Kptfile is a KRM resource and
+follows the familiar structure of KRM resources, the name of the package is also available from the
+`metadata.name` field. This must always be the same of the directory name and kpt will update it
+automatically when forking a package. In this case, `metadata.name` is set `mywordpress`.
```

Comment:
This makes me wonder if end users have to really worry at all about the `metadata.name` ? It seems to be completely internal to `kpt` right ?

Reply:
I don't think we want to force users into a flow that involves a `publish` command, so I think we need a solution that doesn't solely rely on enforcement during publishing.

I do think we want a `validate` command that can handle this and any other issues in the structure/configuration of a package. This would also be useful for users as a git commit hook or as part of a CI pipeline.
The validation can also be part of other commands, we just have to decide if a violation here is fatal or we just provide a warning.

Another alternative would be that we just stop setting the name in the Kptfile altogether as part of the init, get and update commands, and we just say that the name of a package is given by the folder. If users want to set it themselves, they can but we simply ignore it. Forcing them to be the same seemed like a good idea at the time, but maybe it is easier if we just completely ignore the `metadata.name` field.

Reply:
One thing to think about: what other checks would be part of the validation? My intuition is to make `render` the one-stop shop for ensuring package is in a valid and reconciled state, but let's discuss if there's a use case to have a "validate" or "vet" command that's separate.